Lib/imaplib.py

```diff
@@ -1279,6 +1389,251 @@ def print_log(self):
+        except BaseException:
+            imap._idle_capture = False
+            raise
+
+        self._sock_timeout = imap.sock.gettimeout() if imap.sock else None
+        if self._sock_timeout is not None:
+            imap.sock.settimeout(None)  # Socket timeout would break IDLE
+
+        self._old_state = imap.state
+        imap.state = 'IDLING'
+
+        return self
+
+    def __iter__(self):
+        return self
+
+    def _wait(self, timeout=None):
+        # Block until the next read operation should be attempted, either
+        # because data becomes availalable within 'timeout' seconds or
+        # because the OS cannot determine whether data is available.
+        # Return True when a blocking read() is worth trying
+        # Return False if the timeout expires while waiting
+
+        imap = self._imap
+        if timeout is None:
+            return True
+        if imap._readbuf:
+            return True
+        if timeout <= 0:
+            return False
+
+        if imap.sock:
+            fileobj = imap.sock
+        elif platform.system() == 'Windows':
+            return True  # Cannot select(); allow a possibly-blocking read
+        else:
+            fileobj = imap.readfile
+
+        if __debug__ and imap.debug >= 4:
+            imap._mesg(f'idle _wait select({timeout})')
+
+        with selectors.DefaultSelector() as sel:
+            sel.register(fileobj, selectors.EVENT_READ)
+            readables = sel.select(timeout)
+            return bool(readables)
+
+    def _pop(self, timeout, default=('', None)):
+        # Get the next response, or a default value on timeout
+        #
+        # :param timeout:   Time limit (in seconds) to wait for response
+        # :type timeout:    int|float|None
+        # :param default:   Value to return on timeout
+        #
+        # Note: This method ignores 'duration' in favor of the timeout argument.
+        #
+        # Note: Windows IMAP4_stream connections will ignore the timeout
+        # argument and block until the next response arrives, because
+        # Windows select() only works on sockets.
+
+        imap = self._imap
+        if imap.state != 'IDLING':
+            raise imap.error('_pop() only works during IDLE')
+
+        if imap._idle_responses:
+            # Response is ready to return to the user
+            resp = imap._idle_responses.pop(0)
+            if __debug__ and imap.debug >= 4:
+                imap._mesg(f'idle _pop({timeout}) de-queued {resp[0]}')
+            return resp
+
+        if __debug__ and imap.debug >= 4:
+            imap._mesg(f'idle _pop({timeout})')
+
+        if not self._wait(timeout):
+            if __debug__ and imap.debug >= 4:
+                imap._mesg(f'idle _pop({timeout}) done')
+            return default
+
+        if __debug__ and imap.debug >= 4:
+            imap._mesg(f'idle _pop({timeout}) reading')
+        imap._get_response()  # Reads line, calls _append_untagged()
```

Comment:
So IIUC if a (malicious or broken) server ever sends a partial line and then just hangs forever, the client also just hangs forever, uninterrupted by the intended duration. (Not just here but anywhere `_get_response()` is called.)

Is this an acceptable risk? Outside IDLE mode we are also vulnerable to this, but here we have a timeout and a select() call (and a potentially long timeout) that makes me worry more about this.

In modern client code this is generally something we worry about (never trust the server); with old protocols it may be less of an issue (presumably one should only connect to trusted IMAP4 servers).

This is not necessarily something to address, I just want to understand that you are okay with the consequences.